flogger/flogger-testing/src/main/kotlin/com/google/common/flogger/testing/AbstractContextDataProviderSpec.kt

```diff
@@ -0,0 +1,401 @@
+            providerTags.shouldBeEmpty()
+            checkCallbackWasExecuted()
+        }
+
+        @Test
+        public fun `with metadata`() {
+            val (key, value) = FOO to "foo"
+            providerMetadata.shouldBeEmpty()
+            context.newContext()
+                .withMetadata(key, value)
+                .run {
+                    // Should be unique because the key is singleton.
+                    providerMetadata.shouldUniquelyContain(key, value)
+                    markCallbackExecuted()
+                }
+            providerMetadata.shouldBeEmpty()
+            checkCallbackWasExecuted()
+        }
+
+        @Test
+        public fun `with a log level map`() {
+            val defaultLevel = Level.FINE
+            val mapping = "foo.bar" to defaultLevel
+            val levelMap = LogLevelMap.create(mapOf(mapping), defaultLevel)
+            val (logger, level) = mapping
+            dataProvider.isLoggingForced(logger, level).shouldBeFalse()
+            context.newContext()
+                .withLogLevelMap(levelMap)
+                .run {
+                    dataProvider.isLoggingForced(logger, level).shouldBeTrue()
+                    markCallbackExecuted()
+                }
+            dataProvider.isLoggingForced(logger, level).shouldBeFalse()
+            checkCallbackWasExecuted()
+        }
+
+        @Test
+        public fun `with merged tags`() {
+            val (name, value1, value2) = listOf("foo", "bar", "baz")
+            val outerTags = Tags.of(name, value1)
+            providerTags.shouldBeEmpty()
+            context.newContext()
+                .withTags(outerTags)
+                .run {
+                    providerTags shouldBe outerTags.asMap()
+                    val innerTags = Tags.of(name, value2)
+                    val allTags = outerTags.merge(innerTags)
+                    context.newContext()
+                        .withTags(innerTags)
+                        .run {
+                            // Double-check to be sure.
+                            providerTags shouldBe allTags.asMap()
+                            providerTags[name] shouldBe setOf(value1, value2)
+                            markCallbackExecuted()
+                        }
+                    providerTags shouldBe outerTags.asMap()
+                }
+            providerTags.shouldBeEmpty()
+            checkCallbackWasExecuted()
+        }
+
+        /**
+         * Tests a context with concatenated metadata.
+         *
+         * Please note, metadata concatenation is different from metadata merging.
+         * In particular, if during concatenation, several values are discovered
+         * for a singleton key, then it effectively becomes repeated.
+         *
+         * And this aspect is checked by the test.
+         */
+        @Test
+        @Suppress("MagicNumber") // The assertion is readable without a constant.
+        public fun `with concatenated metadata`() {
```

Comment:
Maybe, we could try reducing the amount of code _and_ nesting?

For instance, we could shorten the variable names. And extract some bits into methods?

Please consider.